### PR TITLE
Prompt layer-wise recompute when applicable

### PR DESCRIPTION
**Pull request by pengwa**
### Prompt layer-wise when applicable

Give explicit prompts in export failures to users to enable layer-wise memory optimization if we found the checkpoint function is used.
- Using checkpoint function is a strong indicator that the model is too large to fit in GPU memory.
- If we don't override the checkpoint function here, mostly ONNX export will be failed. 1. For old version PyTorch, when handling gradient checkpoint feature, we just throw an exception. 2. For new version PyTorch, an export failure happens.
- But both failures did not give users explicitly "HOW" to mitigate. This PR did that. 

``

image

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 from onnxruntime.capi._pybind_state import is_ortmodule_available
 
 from ._fallback import ORTModuleFallbackException, ORTModuleInitException, _FallbackPolicy, wrap_exception
+from .options import _MemoryOptimizationLevel
```
Comment (frank-dong-ms) [inline orttraining/orttraining/python/training/ortmodule/__init__.py:18]: This comment is not specific to this line of code.

From your PR description:
1. Using checkpoint function is a strong indicator that the model is too large to fit in GPU memory.
2. But still there is a chance user models use checkpoint function for non-layerwise gradient checkpointing, which we don't think is a common use case.
3. On the other hand, if we don't override the checkpoint function here, mostly ONNX export will be failed.

I agree with your point 1 and 3. 
For 2, yes, non-layerwise checkpoint is not common, but in case some user has this usage, maybe it is better to have a way for user to disable ORT layer-wise-recompute?

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:19]: Agree. I have add following checks: 

`    The actual check is in the overridden function, verifying if:
    1) `checkpoint` is called during ORTModule model export,
    2) Gradient checkpoint autograd function is disallowed (ORTMODULE_ALLOW_AUTOGRAD_CHECKPOINT).
    If true, we reset memory optimization to layer-wise recompute.`

If user explicitly defined `ORTMODULE_MEMORY_OPT_LEVEL`, we will respect it instead of auto enabling it. In this way, they can choose to turn it ON if they found case 2.

File: docs/Memory_Optimizer.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ There are two modes to enable the memory optimizations:
 	```
 3. As shown above, `Config` is a string representative for a re-computable subgraph. All are enabled for recompute in this case.
 4. By `export ORTMODULE_MEMORY_OPT_LEVEL=2`, all plans including compromised recomptable subgraphs will also be enabled.
+5. A side effect specifying `export ORTMODULE_MEMORY_OPT_LEVEL=1/2`, automatically layerwise recompute will be skipped. This is to make sure, user's setting is fully respected.
```
Comment (wschin) [inline docs/Memory_Optimizer.md:55]: 1/2? 0.5? Maybe `1 or 2`?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -6599,3 +6602,146 @@ def run_step(model, attn_weight):
...
+    assert (
+        model._torch_module._execution_manager._training_manager._runtime_options.memory_optimization_level
+        == expected_level_before_run
+    )
+
+    # Forward pass
+    is_export_failed = False
+    try:
+        _ = model(input)
+    except Exception as e:
```
Comment (wschin) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6648]: Consider re-throwing underlying error.

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: This should be caught otherwise the test will crash on some versions of PyTorch . Use a context manager to catch potential failures.

File: docs/Memory_Optimizer.md [hunk 4]
```diff
@@ -30,10 +30,10 @@ Integrate models using `ORTModule`.
...
-- User Specified Subgraph Recompute, enabled by `export ORTMODULE_MEMORY_OPT_LEVEL=0` and `export ORTMODULE_MEMORY_OPT_CONFIG=<plan1 config>,<plan2 config>,...`. This is an advanced usage, that allows users to find the most suitable graphs to recompute, at the cost of overhead to look for the best plans.
+- Transformer layerwise recompute, e.g. aggressively recompute all supported nodes within each transformer layer (usually including attention and mlp sublayers), enabled by `export ORTMODULE_MEMORY_OPT_LEVEL=1`. In this mode, `ORTMODULE_MEMORY_OPT_CONFIG` env values passed by users are not respected.
+- Mamual selected subgraph recompute, enabled by `export ORTMODULE_MEMORY_OPT_LEVEL=0` and `export ORTMODULE_MEMORY_OPT_CONFIG=<plan1 config>,<plan2 config>,...`. This is an advanced usage, that allows users to find the most suitable graphs to recompute, at the cost of overhead to look for the best plans.
```
Comment (wschin) [inline docs/Memory_Optimizer.md:34]: typo: manual

File: orttraining/orttraining/python/training/ortmodule/__init__.py [hunk 5]
```diff
@@ -21,7 +23,10 @@
     raise ImportError("ORTModule is not supported on this platform.")
 
 
-def _defined_from_envvar(name, default_value, warn=True):
+def _defined_from_envvar(name: str, default_value: any, warn: bool = True):
+    """Check given name exists in the environment variable and return the value using the default_value's
+    type if it exists.
+    """
     new_value = os.getenv(name, None)
```
Comment (wschin) [inline orttraining/orttraining/python/training/ortmodule/__init__.py:30]: return os.getenv(name, default_value)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 6]
```diff
@@ -6599,3 +6602,86 @@ def run_step(model, attn_weight):
...
+        except ImportError:
+            # skip if deepspeed is not installed (in amd CI)
+            pass
+    elif fx == "torch":
+        from torch.utils.checkpoint import checkpoint
+
+    original_val = os.environ.get("ORTMODULE_MEMORY_OPT_LEVEL", None)
+    original_val_allow_gradient_checkpoint_export = os.environ.get("ORTMODULE_ALLOW_AUTOGRAD_CHECKPOINT", None)
+
+    if memory_optimization_level is not None:
```
Comment (wschin) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6625]: I think we need to list the supported combinations of these flags and their meanings.
E.g., 
memory_optimization_level=1, allow_gradient_checkpoint_export=0 => does something
memory_optimization_level=2, allow_gradient_checkpoint_export=0 => does something
memory_optimization_level=2, allow_gradient_checkpoint_export=1 => does something

etc..

**Review by wschin — DISMISSED** [review]
Good efforts to consolidates the flags we have.

File: orttraining/orttraining/python/training/ortmodule/__init__.py (hunk 5 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/__init__.py:30]: I think the intention of this function is a bit different, you can check the comment and test case for what it is trying to do. 

`

def test__defined_from_envvar():
    os.environ["DUMMY_ORTMODULE"] = "15"
    assert ortmodule_module._defined_from_envvar("DUMMY_ORTMODULE", 14) == 15
    os.environ["DUMMY_ORTMODULE"] = "15j"
    with warnings.catch_warnings(record=True) as w:
        assert ortmodule_module._defined_from_envvar("DUMMY_ORTMODULE", 14) == 14
        assert len(w) == 1
        assert issubclass(w[-1].category, UserWarning)
        assert "Unable to overwrite constant" in str(w[-1].message)
    del os.environ["DUMMY_ORTMODULE"]
`

orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 6 above)
Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6635]: Added.

**Review by wschin — DISMISSED** [review]
Thanks.

**Review by mindest — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks @wschin @mindest !
